program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,371 @@ use crate::c_oracle_header::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    ///initializes a zero initialized tracker
+    fn initialize(&mut self) -> Result<(), OracleError>;
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError>;
+    /// returns the minimum number of entries that needs to be added
+    /// to the current entry of the time_machine so that the following two
+    /// conditions are satisfied
+    /// 1. after adding the output, the new current entry has the same offset as if
+    ///     the price was constantly updating
+    /// 2. any entry that would have been passed through during this time interval, would be
+    ///     passed through if we skip one entry at a time the number that this outputs
+    fn get_num_skipped_entries_capped(
+        prev_time: i64,
+        current_time: i64,
+    ) -> Result<usize, OracleError> {
+        let time_since_begining_of_current_entry =
+            (prev_time % GRANUALITY) + current_time - prev_time;
+        let mut num_skiped_entries: usize =
+            try_convert(time_since_begining_of_current_entry / GRANUALITY)?;
+        //if num_skipped is greater than NUM_ENTRIES, cap it at the smallest
+        //number greater than NUM_ENTRIES that has the same residue class
+        if num_skiped_entries >= NUM_ENTRIES {
+            num_skiped_entries %= NUM_ENTRIES;
+            num_skiped_entries += NUM_ENTRIES;
+        }
+        Ok(num_skiped_entries)
+    }
+    ///gets the index of the next entry
+    fn get_next_entry(current_entry: usize) -> usize {
+        (current_entry + 1) % NUM_ENTRIES
+    }
+
+    ///returns the remining time before the next multiple of GRANUALITY
+    fn get_time_to_entry_end(time: i64) -> i64 {
+        (GRANUALITY - (time % GRANUALITY)) % GRANUALITY
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an SMA Tracker that has NUM_ENTRIES entries
+/// each tracking time weighted sums for GRANUALITY seconds periods.
+///The prices are assumed to be provided under some fixed point representation, and the computation
+/// gurantees accuracy up to the last decimal digit in the fixed point representation.
+/// Assumes THRESHOLD < GRANUALITY
+pub struct SmaTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    valid_entry_counter:        u64, // is incremented everytime we add a valid entry
```

Comment:
What is this being used for?

Reply:
I don't get how this works -- it seems like valid_entry_counter continuously increments and never resets. How do i use that to determine that the intervening entries are valid?

Reply:
the thing I'm confused about is that `valid_entry_counter` is a purely incrementing thing. I would have expected to have a count stored per entry or something such that you can subtract?

Reply:
Oh I understand the confusion now, so, I did not see a reason to store it per entry because we can just count the number of valid entries with in the last  NUM_entries, it was primarily here for when we get two different VAAs. We can use it to infer the Number of Valid Entries before each entry.

Though storing it for each entry might make the design easier, if it is not too expensive, am happy to do either.